File: crates/wasm/examples/decrypt.ts
#!/usr/bin/env -S deno run --allow-read --allow-write

// SPDX-FileCopyrightText: 2024 Shun Sakai
//
// SPDX-License-Identifier: Apache-2.0 OR MIT

import * as cli from "@std/cli";
import * as io from "@std/io";

import * as command from "@cliffy/command";

import * as abcrypt from "../pkg/abcrypt_wasm.js";

import { VERSION } from "./version.ts";

const { args, options } = await new command.Command()
  .name("decrypt")
  .version(VERSION)
  .description(
    "An example of decrypting from the abcrypt encrypted data format.",
  )
  .option("-o, --output <FILE:file>", "Output the result to a file.")
  .arguments("<FILE:file>")
  .parse();

const ciphertext = Deno.readFileSync(args[0]);

const passphrase = new TextEncoder()
  .encode(cli.promptSecret("Enter passphrase: ")!);
const plaintext = abcrypt.decrypt(ciphertext, passphrase);

if (options.output === undefined) {
  io.writeAllSync(Deno.stdout, plaintext);
} else {
  Deno.writeFileSync(options.output, plaintext);
}

File: crates/python/examples/decrypt.py
#!/usr/bin/env python3

# SPDX-FileCopyrightText: 2024 Shun Sakai
#
# SPDX-License-Identifier: Apache-2.0 OR MIT

import argparse
import getpass
import sys

import abcrypt_py
import version


def main() -> None:
    parser = argparse.ArgumentParser(
        prog="decrypt",
        description="An example of decrypting a file from the abcrypt encrypted data format",
    )
    parser.add_argument(
        "-V", "--version", action="version", version=version.VERSION
    )
    parser.add_argument(
        "-o",
        "--output",
        default=sys.stdout.buffer,
        type=argparse.FileType("wb"),
        help="output the result to a file",
        metavar="FILE",
    )
    parser.add_argument(
        "input",
        nargs="?",
        default=sys.stdin,
        type=argparse.FileType("rb"),
        help="input file",
        metavar="FILE",
    )
    args = parser.parse_args()

    ciphertext = args.input.read()

    passphrase = bytes(getpass.getpass("Enter passphrase: "), encoding="utf-8")
    plaintext = abcrypt_py.decrypt(ciphertext, passphrase)

    args.output.write(plaintext)


if __name__ == "__main__":
    main()
